src/metabase/query_processor/streaming/xlsx.clj

```diff
@@ -392,25 +393,58 @@
         (str column-title " (" (currency-identifier col-viz-settings) ")")
         column-title))))
 
+(def ^:private auto-sizing-threshold
+  "The number of rows used for auto-sizing columns. If this number is too large, exports of large datasets
+  will be prohibitively slow."
+  1000)
+
+(def ^:private extra-column-width
+  "The extra width applied to columns after they have been auto-sized, in units of 1/256 of a character width.
+  This ensures the cells in the header row have enough room for the filter dropdown icon."
+  (* 4 256))
+
+(defn- autosize-columns!
+  "Adjusts each column to fit its largest value, plus a constant amount of extra padding."
+  [sheet col-count]
+  (doseq [i (range col-count)]
+    (.autoSizeColumn ^SXSSFSheet sheet i)
+    (.setColumnWidth ^SXSSFSheet sheet i (+ (.getColumnWidth ^SXSSFSheet sheet i) extra-column-width)))
+  (.untrackAllColumnsForAutoSizing ^SXSSFSheet sheet))
+
+(defn- setup-header-row!
+  "Turns on auto-filter for the header row, which adds a button to each header cell that allows columns to be
+  filtered and sorted. Also freezes the header row so that it floats above the data."
+  [sheet col-count]
+  (when (> col-count 0)
+    (.setAutoFilter ^SXSSFSheet sheet (new CellRangeAddress 0 0 0 (dec col-count)))
+    (.createFreezePane ^SXSSFSheet sheet 0 1)))
+
 (defmethod i/streaming-results-writer :xlsx
   [_ ^OutputStream os]
   (let [workbook            (SXSSFWorkbook.)
         sheet               (spreadsheet/add-sheet! workbook (tru "Query result"))]
     (reify i/StreamingResultsWriter
       (begin! [_ {{:keys [ordered-cols]} :data} {col-settings ::mb.viz/column-settings}]
+        (.trackAllColumnsForAutoSizing ^SXSSFSheet sheet)
+        (setup-header-row! sheet (count ordered-cols))
         (spreadsheet/add-row! sheet (column-titles ordered-cols col-settings)))
 
-      (write-row! [_ row _ ordered-cols {:keys [output-order] :as viz-settings}]
+      (write-row! [_ row row-count ordered-cols {:keys [output-order] :as viz-settings}]
```

Comment:
Minor suggestion: I think it would be clearer if you named this `row-num` instead of `row-count`. I had to scratch my head a little and try to figure out _how_ we could possibly know the total `row-count` at any given moment since we're streaming the rows as we get them... 

After I looked at this a bit more the variable is called `row-count` in the reducing function (`metabase.query-processor.streamingmetabase.query-processor.streaming/streaming-rff`) as well. It is actually a counter in that situation altho I think maybe we ought to change the name to `row-num` there as well. FWIW it _is_ called `row-num` in `metabase.query-processor.streaming.interface`.